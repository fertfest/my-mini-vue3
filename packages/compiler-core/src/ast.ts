import { CREATE_ELEMENT_VNODE } from "./runtimeHelper";

export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION,
}

export const enum ElementTypes {
  ELEMENT,
}

export function createVNodeCall(context, tag, props, children) {
  const { addHelper } = context;
  addHelper(CREATE_ELEMENT_VNODE);
  return {
    type: NodeTypes.ELEMENT,
    tag,
    props,
    children,
  };
}
