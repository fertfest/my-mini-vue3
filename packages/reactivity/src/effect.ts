import { extend } from "@my-mini-vue3/shared";

export class ReactiveEffect {
  private _fn: any;
  private _firstRun: boolean = true;
  private _effectsArray: any[] = [];
  onStop?: () => void;
  public scheduler?: () => void;

  constructor(fn, scheduler?) {
    this._fn = fn;
    if (scheduler) {
      this.scheduler = scheduler;
    }
  }

  pushEffects(effects) {
    this._effectsArray.push(effects);
  }

  stop() {
    if (this.onStop) {
      this.onStop();
    }

    if (this._effectsArray.length === 0) {
      return;
    }

    // 对于每一个 effects, 都从 effects 中清除当前 effect(this)
    for (const effects of this._effectsArray) {
      effects.delete(this);
    }

    this._effectsArray = [];
  }

  run() {
    // 在执行功能函数之前设置当前 ReactiveEffect 为 activeEffect
    activeEffect = this;
    return this._fn();
  }
}

export let activeEffect: any = undefined;

export function setActiveEffect(effect: any) {
  activeEffect = effect;
}

/**
 * 接收一个函数 fn 并执行它, 在首次执行时, 函数内部被访问的响应式变量都会 track 包装该函数的 effect(ReactiveEffect).
 * 这样, 当响应式变量的值更改并触发依赖时, ReactiveEffect 内部的 fn 会被再次执行.
 * @param fn 一个函数
 * @param options 可能包含 scheduler, onStop ...
 * @returns
 */
export function effect(fn, options?: any) {
  const _effect = new ReactiveEffect(fn);

  extend(_effect, options);

  _effect.run();
  const res = fn.bind(_effect);
  res.effect = _effect;
  return res;
}

export function stop(runner: any) {
  const _effect = runner.effect;
  _effect.stop();
}

const targetMap = new Map<any, any>();
export function track(target, key) {
  if (!activeEffect) {
    return;
  }

  // 先通过 targetMap.get(target) 找到 Map<key, effects>
  let keyMap: Map<any, any> = targetMap.get(target);
  if (!keyMap) {
    // target 对应的 map 还没有被创建, 需要创建
    keyMap = new Map();

    // 创建后需要将对应的 map 添加到targetMap中
    targetMap.set(target, keyMap);
  }

  let effects: Set<any> = keyMap.get(key);
  if (!effects) {
    // 同样如果 key 对应的 set 不存在, 需要创建
    effects = new Set();
    keyMap.set(key, effects);
  }

  trackActiveEffect(effects);

  activeEffect = undefined;
}

export function trackActiveEffect(dep) {
  if (!isTracking()) {
    return;
  }
  // 防止重复添加
  if (dep.has(activeEffect)) return;

  dep.add(activeEffect);
  activeEffect.pushEffects(dep);
}

export function trigger(target, key) {
  if (!target || !key) {
    return;
  }

  const keyMap = targetMap.get(target);
  if (!keyMap) {
    return;
  }

  const effects = keyMap.get(key);
  if (!effects) {
    return;
  }
  triggerEffects(effects);
}

export function triggerEffects(effects) {
  for (const effect of effects) {
    if (effect.scheduler) {
      effect.scheduler();
    } else {
      effect.run();
    }
  }
}

export function isTracking() {
  return activeEffect !== undefined;
}
