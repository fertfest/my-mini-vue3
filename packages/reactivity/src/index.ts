export { ref, proxyRefs, isRef, unRef } from "./ref";
export {
  reactive,
  readonly,
  shallowReadonly,
  isReactive,
  isProxy,
} from "./reactive";
export { effect } from "./effect";
