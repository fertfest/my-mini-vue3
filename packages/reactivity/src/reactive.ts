import { isObject } from "@my-mini-vue3/shared";
import {
  mutableHandlers,
  readonlyHandlers,
  shallowReadonlyHandlers,
} from "./baseHandlers";

export const enum ReactiveFlags {
  IS_REACTIVE = "__v_isReactive",
  IS_READONLY = "__v_isReadonly",
}

export function reactive(target) {
  return createReactiveObject(target, mutableHandlers);
}

export function readonly(target) {
  return createReactiveObject(target, readonlyHandlers);
}

export function shallowReadonly(target) {
  return createReactiveObject(target, shallowReadonlyHandlers);
}

export function isReactive(value) {
  return !!value[ReactiveFlags.IS_REACTIVE];
}

export function isReadonly(value) {
  return !!value[ReactiveFlags.IS_READONLY];
}

export function isProxy(value) {
  return isReactive(value) || isReadonly(value);
}

function createReactiveObject(target: any, baseHandlers) {
  if (!isObject(target)) {
    console.warn(`target ${target} 必须是一个对象`);
    return target;
  }
  return new Proxy(target, baseHandlers);
}
