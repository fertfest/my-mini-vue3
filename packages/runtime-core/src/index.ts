export { h } from "./h";
export { renderSlot } from "./helpers/renderSlot";
export { createTextVNode, createElementVNode } from "./vnode";
export { getCurrentInstance, registerRuntimeCompiler } from "./component";
export { inject, provide } from "./apiInject";
export { createRenderer } from "./renderer";
export { queueJobs, nextTick } from "./scheduler";
export { toDisplayString } from "@my-mini-vue3/shared";
export * from "@my-mini-vue3/reactivity";
