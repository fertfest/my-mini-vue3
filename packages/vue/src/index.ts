export * from "@my-mini-vue3/runtime-dom";

import { baseCompile } from "@my-mini-vue3/compiler-core";
import * as runtimeDom from "@my-mini-vue3/runtime-dom";
import { registerRuntimeCompiler } from "@my-mini-vue3/runtime-dom";

function compileToFunction(template) {
  const { code } = baseCompile(template);
  const render = new Function("Vue", code)(runtimeDom);

  return render;
}

registerRuntimeCompiler(compileToFunction);
