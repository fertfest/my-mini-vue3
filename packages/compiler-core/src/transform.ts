import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelper";

export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createRootCodegen(root);
  root.helpers = [...context.helpers];
}

function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms ?? [],
    helpers: new Set(),
    addHelper(newHelper) {
      context.helpers.add(newHelper);
    },
  };

  return context;
}
function traverseNode(node: any, context: any) {
  const { nodeTransforms, addHelper } = context;
  const exitFns: any[] = [];
  for (const t of nodeTransforms) {
    const onExit = t(node, context);
    if (onExit) exitFns.push(onExit);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      addHelper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
      break;
  }

  for (const fn of [...exitFns].reverse()) {
    fn();
  }
}
function traverseChildren(node: any, context: any) {
  const children = node.children;
  if (children) {
    for (const child of children) {
      traverseNode(child, context);
    }
  }
}
function createRootCodegen(root: any) {
  const child = root.children[0];
  if (child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = child;
  }
}
