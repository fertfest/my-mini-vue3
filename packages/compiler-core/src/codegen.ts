import { isString } from "@my-mini-vue3/shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  TO_DISPLAY_STRING,
  helperNameMap,
} from "./runtimeHelper";
import { transform } from "./transform";

export function generate(ast) {
  const context = createGenerateContext(ast);
  const { push } = context;

  genFunctionPreamble(ast, context);

  push("return ");
  const args = ["_ctx", "_cache"];
  const functionName = "render";
  const signature = `${functionName}(${args.join(",")})`;

  push("function " + signature + "{ return ");

  transform(ast);
  genNode(ast.codegenNode, context);

  push(" }");

  return {
    code: context.code,
  };
}
function genFunctionPreamble(ast: any, context: any) {
  const { push } = context;
  if (ast.helpers.length > 0) {
    const VueBinging = "Vue\n";
    const aliasHelper = (s) => `${helperNameMap[s]}: _${helperNameMap[s]}`;
    push(
      `const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${VueBinging}`
    );
  }
}

function genNode(node: any, context: any) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
  }
}
function genText(node: any, context: any) {
  const { push } = context;
  push(`'${node.content}'`);
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}

function genElement(node, context) {
  const { push, helper } = context;
  const { tag, children, props } = node;

  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  genNodeList(genNullable([tag, props, children]), context);
  // genNode(children, context);

  push(")");
}

function genNodeList(nodes: any[], context) {
  const { push } = context;

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }

    if (i !== nodes.length - 1) {
      push(", ");
    }
  }
}

function genNullable(args: any[]) {
  return args.map((arg) => arg || "null");
}

function createGenerateContext(ast: any) {
  const context = {
    code: "",
    push(str: string) {
      context.code += str;
    },
    helper(key) {
      return `_${helperNameMap[key]}`;
    },
  };

  return context;
}
function genCompoundExpression(node: any, context: any) {
  const { push } = context;
  for (const child of node.children) {
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}
