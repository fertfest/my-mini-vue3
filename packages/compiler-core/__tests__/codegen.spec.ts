import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformElement } from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformText } from "../src/transforms/transformText";

// test("codegen", () => {
//   const ast = baseParse("hi 1");

//   const { code } = generate(ast);

//   expect(code).toMatchSnapshot();
// });

// test("interpolation", () => {
//   const ast = baseParse("{{message}}");
//   console.log("ast", ast);

//   transform(ast, {
//     nodeTransforms: [transformExpression],
//   });
//   const { code } = generate(ast);
//   expect(code).toMatchSnapshot();
// });

// test("interpolation module", () => {
//   const ast = baseParse("{{hello}}");
//   transform(ast, {
//     nodeTransforms: [transformExpression],
//   });

//   const { code } = generate(ast);
//   expect(code).toMatchSnapshot();
// });

test("element", () => {
  const ast: any = baseParse("<div>hi, {{message}}</div>");
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });

  const { code } = generate(ast);
  expect(code).toMatchSnapshot();
});

// test("element and interpolation", () => {
//   const ast = baseParse("<div>hi,{{msg}}</div>");
//   transform(ast, {
//     nodeTransforms: [transformElement, transformText, transformExpression],
//   });

//   const { code } = generate(ast);
//   expect(code).toMatchSnapshot();
// });
