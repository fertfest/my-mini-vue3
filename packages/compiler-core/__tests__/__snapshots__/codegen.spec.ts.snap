// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`element 1`] = `
"const { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } = Vue
return function render(_ctx,_cache){ return _createElementVNode('div', null, 'hi, ' + _toDisplayString(_ctx.message)) }"
`;
